src/mysql-util/src/privileges.rs

```diff
@@ -0,0 +1,241 @@
+                                .iter()
+                                .any(|privs| {
+                                    if let Some(privs) = privs {
+                                        if privs.contains(*privilege)
+                                            || privs.contains("ALL PRIVILEGES")
+                                        {
+                                            return true;
+                                        }
+                                    }
+                                    false
+                                })
+                        } else {
+                            false
+                        }
+                    })
+                {
+                    // This privilege was not found for this table
+                    Some(MissingPrivilege {
+                        privilege: privilege.to_string(),
+                        qualified_table_name: format!("{}.{}", table.schema_name, table.table_name),
+                    })
+                } else {
+                    None
+                }
+            })
+        })
+        .collect::<Vec<_>>();
+
+    // Check that we have REPLICATION SLAVE priviliges at the global level
+    if !grant_map
+        .get("*")
+        .and_then(|schema_map| schema_map.get("*"))
+        .map(|privs| privs.contains("REPLICATION SLAVE") || privs.contains("ALL PRIVILEGES"))
+        .unwrap_or(false)
+    {
+        errors.push(MissingPrivilege {
+            privilege: "REPLICATION SLAVE".to_string(),
+            qualified_table_name: "*.*".to_string(),
+        });
+    }
+
+    if !errors.is_empty() {
+        Err(MySqlError::MissingPrivileges(errors))
+    } else {
+        Ok(())
+    }
+}
+
+/// Regex copied and adjusted from https://github.com/ansible/ansible/pull/40092/
+/// If this regex matches then this is a grant on an actual object which looks like:
+///     GRANT SELECT, INSERT, UPDATE ON `db1`.* TO `u1`@`localhost`
+///     GRANT SELECT ON `db1`.`table1` TO `my_user`@`localhost` WITH GRANT OPTION
+/// The regex was tested on https://regexr.com/ and in the tests below. It needs to account
+/// for multiple quoting styles, the possibility of a wildcard schema or table, and for the
+/// quote-char to be part of the table/schema name too.
+/// Group 1 is the list of privileges being granted
+/// Group 2 is either the wildcard * or a quoted database/schema
+/// Group 4 is the unquoted database/schema when the wildcard is not matched
+/// Group 5 is either the wildcard * or a quoted table
+/// Group 7 is the unquoted table when the wildcard is not matched
+/// Group 9 is the user being granted
+/// We use the fancy_regex crate to allow backreferences which are necessary to find the ending
+/// quote of each identifier since there are different quoting types across mysql versions.
+static GRANT_REGEX: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r#"GRANT (.+) ON (\*|(['`"])(.*)\3).(\*|(['`"])(.*)\6) TO (['`"])(.*)\8@.*"#)
```

Comment:
I spent a lot of time iterating-on and testing this regex -- happy to talk through it live if the comments / test cases aren't clear enough.

You can also play with it easily on [regexr.com](https://regexr.com/)

and use these tests 
```
GRANT SELECT, INSERT, UPDATE ON `db1`.* TO `u1`@`localhost` IDENTIFIED BY PASSWORD `test`
GRANT USAGE ON *.* TO `u1`@`localhost`  
GRANT SUPER, CREATE TEMPORARY TABLES, LOCK TABLES ON 'db1'.'table1' TO `u1`@`localhost`
GRANT SUPER, CREATE TEMPORARY TABLES, LOCK TABLES ON *.`rw` TO `u1`@`localhost`
GRANT SUPER, CREATE TEMPORARY TABLES, LOCK TABLES ON *.`r`w` TO `u1`@`localhost`
GRANT SELECT, INSERT, UPDATE ON `таблица`.'mixED_CAse' TO `u1`@`localhost`
GRANT SELECT, INSERT, UPDATE ON `db1`.* TO `u1`@`localhost`
```